src/Components/Components/src/Routing/Router.cs

```diff
@@ -58,10 +56,15 @@ public class Router : IComponent, IHandleAfterRender, IDisposable
         /// </summary>
         [Parameter] public RenderFragment<RouteData> Found { get; set; }
 
+        /// <summary>
+        /// Get or sets the content to display when the upcoming route is loading.
+        /// </summary>
+        [Parameter] public RenderFragment Loading { get; set; }
+
         /// <summary>
         /// Gets or sets a handler that should be called before navigating to a new page.
         /// </summary>
-        [Parameter] public Action<string> OnNavigate { get; set; }
+        [Parameter] public Func<string, Task<bool>> OnNavigateAsync { get; set; }
```

Comment:
For this callback, instead of passing a single `string` argument, what about possibly defining a new type like:

```
public class NavigationEventArgs
{
    public string Url { get; }
}
```

Reason: we might well want to add more info in the future, and having a class means we're free to do so non-breakingly. If we only pass a string, we can't add more params in the future.

Also we could use it to generalise the "knowing when to refresh assemblies" part of this process. Returning a bool is slightly nonobvious (I didn't know what the bool meant until reading the sources). There could be another property on `NavigationEventArgs` that you can use to trigger the re-loading. For example it could be an `Action` you can invoke, or it could be an instance of some new `IRouteTable` that has an API for updating it. Then people using `OnNavigateAsync` for reasons unrelated to lazy loading can completely ignore that and don't have to know about any return value.

Reply:
I debated using the existing `LocationChangeEventArgs` object for this but opted against it since that type exposes the absolute path, and not the relative one. Your point about adding a more explicit strategy for allowing the user to update the route table makes sense to.

> or even AddRoutesFromAssembly(assembly) that's additive on top of the existing data (though I know that's more complicated - I'm not saying we should do all this - just listing thoughts for consideration!).

Indeed it is. I took a stab at this in one of the commits. Of course, ensure that the route table always has the right precedence sorting gets wonky when you support an additive API.

Of the options, a `RefreshRoutes` callback seems like the most straightforward option.

> Then people using OnNavigateAsync for reasons unrelated to lazy loading can completely ignore that and don't have to know about any return value.

I'm in agreement with you there. I was working towards a point where the API interface for the method would be `async Task OnNavigateAsync(string uri)`.